src/DotNetty.Transport/Channels/Sockets/DatagramChannel.cs

```diff
@@ -0,0 +1,460 @@
+        {
+        }
+
+        public DatagramChannel(Socket socket)
+            : base(null, socket)
+        {
+            this.config = new DefaultDatagramChannelConfig(this, socket);
+
+            // Read is not allowed until socket is bound for UDP
+            this.config.AutoRead = false;
+        }
+
+        public override IChannelConfiguration Configuration => this.config;
+
+        public override ChannelMetadata Metadata => ChannelMetadata;
+
+        protected override EndPoint LocalAddressInternal => this.Socket.LocalEndPoint;
+
+        protected override EndPoint RemoteAddressInternal => this.Socket.RemoteEndPoint;
+
+        protected override void DoBind(EndPoint localAddress)
+        {
+            this.Socket.Bind(localAddress);
+            this.CacheLocalAddress();
+
+            this.SetState(StateFlags.Active);
+            this.config.AutoRead = true;
+        }
+
+        public override bool Active => 
+            this.Open
+            && (this.config.ActiveOnOpen && this.Registered || this.Socket.IsBound);
+
+        protected override bool DoConnect(EndPoint remoteAddress, EndPoint localAddress)
+        {
+            if (localAddress != null)
+            {
+                this.DoBind(localAddress);
+            }
+
+            bool success = false;
+            try
+            {
+                this.Socket.Connect(remoteAddress);
+                success = true;
+                return true;
+            }
+            finally
+            {
+                if (!success)
+                {
+                    this.DoClose();
+                }
+            }
+        }
+
+        protected override void DoFinishConnect(SocketChannelAsyncOperation operation)
+        {
+            throw new NotSupportedException();
+        }
+
+        protected override void DoDisconnect() => this.DoClose();
+
+        protected override void DoClose()
+        {
+            if (this.TryResetState(StateFlags.Open | StateFlags.Active))
+            {
+                this.Socket.Shutdown(SocketShutdown.Both);
+                this.Socket.Dispose();
+            }
+        }
+
+        protected override void ScheduleSocketRead()
+        {
+            SocketChannelAsyncOperation operation = this.ReadOperation;
+            operation.RemoteEndPoint = new IPEndPoint(
+                this.LocalAddress.AddressFamily == AddressFamily.InterNetwork ? IPAddress.Any : IPAddress.IPv6Any,
+                IPEndPoint.MinPort);
+
+            IRecvByteBufAllocatorHandle handle = this.Unsafe.RecvBufAllocHandle;
+            IByteBuffer buffer = handle.Allocate(this.config.Allocator);
+            handle.AttemptedBytesRead = buffer.WritableBytes;
+            operation.UserToken = buffer;
+
+            ArraySegment<byte> bytes = buffer.GetIoBuffer();
+            operation.SetBuffer(bytes.Array, bytes.Offset, buffer.WritableBytes);
+            bool pending = this.Socket.ReceiveFromAsync(operation);
```

Comment:
I guess it can benefit from `ExecutionContext.SuppressFlow()`. Pls check `AbstractSocketByteChannel`'s version for details. That avoids execution context copy on `ReceiveAsync` call.